stdlib/public/Concurrency/AsyncStreamBackpressure.swift

```diff
@@ -0,0 +1,1961 @@
+                }
+                onTermination?()
+
+            case .none:
+                break
+            }
+        }
+    }
+}
+
+/// The state machine of the backpressured async stream.
+@available(SwiftStdlib 5.1, *)
+struct _StateMachine<Element, Failure: Error> {
+    enum _State {
+        struct Initial {
+            /// The backpressure strategy.
+            var backPressureStrategy: _InternalBackPressureStrategy
+            /// Indicates if the iterator was initialized.
+            var iteratorInitialized: Bool
+            /// The onTermination callback.
+            var onTermination: (@Sendable () -> Void)?
+        }
+
+        struct Streaming {
+            /// The backpressure strategy.
+            var backPressureStrategy: _InternalBackPressureStrategy
+            /// Indicates if the iterator was initialized.
+            var iteratorInitialized: Bool
+            /// The onTermination callback.
+            var onTermination: (@Sendable () -> Void)?
+            /// The buffer of elements.
+            var buffer: _Deque<Element>
+            /// The optional consumer continuation.
+            var consumerContinuation: CheckedContinuation<Element?, Error>?
+            /// The producer continuations.
+            var producerContinuations: _Deque<(UInt, (Result<Void, Error>) -> Void)>
+            /// The producers that have been cancelled.
+            var cancelledAsyncProducers: _Deque<UInt>
+            /// Indicates if we currently have outstanding demand.
+            var hasOutstandingDemand: Bool
+        }
+
+        struct SourceFinished {
+            /// Indicates if the iterator was initialized.
+            var iteratorInitialized: Bool
+            /// The buffer of elements.
+            var buffer: _Deque<Element>
+            /// The failure that should be thrown after the last element has been consumed.
+            var failure: Failure?
+            /// The onTermination callback.
+            var onTermination: (@Sendable () -> Void)?
+        }
+
+        case initial(Initial)
+        /// The state once either any element was yielded or `next()` was called.
+        case streaming(Streaming)
+        /// The state once the underlying source signalled that it is finished.
+        case sourceFinished(SourceFinished)
+
+        /// The state once there can be no outstanding demand. This can happen if:
+        /// 1. The iterator was deinited
+        /// 2. The underlying source finished and all buffered elements have been consumed
+        case finished(iteratorInitialized: Bool)
+
+        /// An intermediate state to avoid CoWs.
+        case modify
+    }
+
+    /// The state machine's current state.
+    var _state: _State
+
+    // The ID used for the next CallbackToken.
+    var nextCallbackTokenID: UInt = 0
+
+    var _onTermination: (@Sendable () -> Void)? {
+        set {
+            switch self._state {
```

Comment:
Good point. I went through all the state switches again and added a move to `modify` in every case where we have a `(var state)`. That is pretty nice because the `var` makes it apparent that we are actually modifying something.